overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,125 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
 	return gadgetInfo.Connections, nil
 }
 
+// deriveNeededSnapRevisions returns revisions that need to be kept for the given snap.
+//
+// A revision is needed only if:
+//    1. Revision is in a snap's sequence
+//    2. Revision is in a snap's refresh hint
+//    3. Revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+func deriveNeededSnapRevisions(st *state.State, snapName string) ([]snap.Revision, error) {
+	var snapst SnapState
+	err := Get(st, snapName, &snapst)
+	// nothing to clean
+	if errors.Is(err, state.ErrNoState) {
+		return []snap.Revision{}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	neededSnapDownloads := []snap.Revision{}
+
+	// keep revisions in snap's sequence
+	for _, seq := range snapst.Sequence {
+		neededSnapDownloads = append(neededSnapDownloads, seq.Revision)
+	}
+	// keep revision in refresh hint (if it exists)
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	} else if hint, ok := refreshHints[snapName]; ok {
+		neededSnapDownloads = append(neededSnapDownloads, hint.Revision())
+	}
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			if snapsup.InstanceName() != snapName {
+				continue
+			}
+			neededSnapDownloads = append(neededSnapDownloads, snapsup.Revision())
+		}
+	}
+
+	return neededSnapDownloads, nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// A download is kept only if:
+//    1. Revision is in a snap's sequence
+//    2. Revision is in a snap's refresh hint
+//    3. Revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
```

Comment:
This function and the one below share some code. Calling `cleanSnapDownloads` here repeatedly would probably be too inefficient but we could either refactor it or split out a small helper to build the map

Reply:
I came up with another approach that is more efficient for `cleanDownloads` but adds very little temp extra memory for `cleanSnapDownloads`. PR updated!